src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationKeyComparer.cs

```diff
@@ -29,49 +29,94 @@ public class ConfigurationKeyComparer : IComparer<string>
         /// <returns>Less than 0 if x is less than y, 0 if x is equal to y and greater than 0 if x is greater than y.</returns>
         public int Compare(string? x, string? y)
         {
-            string[] xParts = x?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-            string[] yParts = y?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-
-            // Compare each part until we get two parts that are not equal
-            for (int i = 0; i < Math.Min(xParts.Length, yParts.Length); i++)
+            if (x is not null && y is not null)
             {
-                x = xParts[i];
-                y = yParts[i];
+                if (x.Contains(ConfigurationPath.KeyDelimiter) && y.Contains(ConfigurationPath.KeyDelimiter))
+                {
+                    int xIndex = 0;
```

Comment:
I find it easier if instead I use the ReadOnlySpan to "remember" the position where I am parsing instead of using an index.

So instead, you can declare a `xSpan` up here being the whole string and the `while` loop condition can be `!xSpan.IsEmpty && !ySpan.IsEmpty`. The `IndexOf` line just becomes `xSpan.IndexOf(ConfigurationPath.KeyDelimiter)`, and then you don't need to subtract off the `xIndex` in your slices below. The slices become `xSpan.Slice(0, nextXIndex)`. And then at the end, move your spans forward by `xSpan = xSpan.Slice(nextXIndex + 1)`.

Here's an example of what I'm talking about:

[https://github.com/xamarin/xamarin-android/blob/1ff560542f1bac8c7a6d5f4437711917eb4d16d8/src/Mono.Android/Android.Runtime/AndroidRuntime.cs#L474-L485](https://github.com/xamarin/xamarin-android/blob/1ff560542f1bac8c7a6d5f4437711917eb4d16d8/src/Mono.Android/Android.Runtime/AndroidRuntime.cs#L474-L485)